rai/node/wallet.cpp

```diff
@@ -925,43 +929,73 @@ std::shared_ptr<rai::block> rai::wallet::change_action (rai::account const & sou
 	return block;
 }
 
-std::shared_ptr<rai::block> rai::wallet::send_action (rai::account const & source_a, rai::account const & account_a, rai::uint128_t const & amount_a, bool generate_work_a)
+std::shared_ptr<rai::block> rai::wallet::send_action (rai::account const & source_a, rai::account const & account_a, rai::uint128_t const & amount_a, bool generate_work_a, boost::optional<std::string> id_a)
 {
 	std::shared_ptr<rai::block> block;
+	bool cached_block = false;
 	{
-		rai::transaction transaction (store.environment, nullptr, false);
-		if (store.valid_password (transaction))
+		rai::transaction transaction (store.environment, nullptr, (bool)id_a);
```

Comment:
Why does this need to be a write transaction if id_a is set?

Reply:
It updates the `send_action_ids` store. Should I add a comment?

Reply:
It's better to only hold a write transaction for the exact amount of time you need it, since it blocks all other readers. You're not even sure you need it at this point.

Reply:
(the send queue is serialized, so there's no race)

Reply:
I'm assuming it's better to release the old transaction before creating the write transaction, instead of using a nested transaction?

Reply:
Yes, annoyingly enough.